oeps/oep-0049-django-app-patterns.rst

```diff
@@ -0,0 +1,165 @@
+
+      def ready(self):
+          """
+          Connect handlers to recalculate grades.
+          """
+          from .signals import handlers
+
+.. _api.py:
+api.py
+++++++
+This should be single point of entry for other Python code to talk to your app. This is *not* a Rest API, this is a Python API (see rest_api_). Some rules for ``api.py`` are as follows:
+
+1. API methods defined in ``api.py`` should be well-named, self-consistent, and relevant to its own domain (without exposing technical and implementation details)
+2. An app's Django models and other internal data structures should not be exposed via its Python APIs.
+3. Ideally, tests should use only Python APIs declared in other apps' ``api.py`` files. However, if an app's API is needed *only* for testing, then test-relevant Python APIs should be defined/exported in an intentional Python module called ``api_for_tests.py``.
+
+
+Not exposing an app's data structures can be tricky because it's very easy to expose them without meaning to. Therefore there are a couple common strategies we employ.
+
+1. When importing internal app code to be used in the ``api.py`` file, prefix it with an underscore so it's clear it's for internal use only.
+
+2. Create a ``data.py`` file to house simple data objects that can be passed from your app's function to the calling app. By creating these objects, we can avoid both passing Django model objects or querysets directly and having to serialize data. Other apps may import data classes from ``data.py`` in additional to functionality from ``api.py``. See data.py_ for more details.
+
+For example:
+
+.. code-block:: python
+
+    from django.conf.settings import UNSUPPORTED_PROGRAM_UUIDS
+
+    from .data import ProgramData
+    from .models_api import get_programs as _get_programs
+
+    def get_supported_programs():
+        """
+        Gets all programs that aren't in UNSUPPORTED_PROGRAM_UUIDS settings
+        """
+        all_programs = _get_programs() # returns queryset
+        supported_programs [
+          ProgramData(
+            uuid=program.uuid,
+            title=program.title,
+            status=program.status
+          )
+          for program in all_programs
+          if program.uuid not in UNSUPPORTED_PROGRAM_UUIDS
+        ]
+        return supported_programs
+
+.. _data.py:
+data.py
++++++++
+This file should include the public data structures for the app that can be passed between apps without exposing internal features. These should be used instead of sending Django model objects or querysets to apps that call the functions in ``api.py``. This file should not import anything other than stdlib modules, so that it may be imported by any other app without issue. These data objects should be simple objects with all business logic handled by ``api.py``. They may however perform simple validation, as long as it is self-contained (doesn't reach out to database, network, or any code outside of the class)
```

Comment:
I think it's slightly better to just recommend `attrs` as the default, so that people don't go calling methods that they expect to be present. But honestly, the shape of the API is so similar that the number of times it's going to bite people is probably pretty small, so I don't feel strongly about it.

Reply:
I'd prefer we pick one because that will make things consistent which in turn will speed up reading.  I would say `attrs` should be the best practice and you can fall to `dataclass` if you have a good reason(where you like it better is not a good reason).

Reply:
The majority wins on this one. I removed the `dataclass` reference.